### PR TITLE
Address Comments from Demo 1

### PR DESCRIPTION
**Pull request by gabrielleungkh**
- Changed error checking within the builder classes so that they don't throw exceptions, but are handled gracefully so no crashes occur
- Made ConcreteMediator and CarrierManagementSystemDirector singletons have synchronized getInstance methods
- In UI, instead of having enum class for defining the number of RF ports for LTE/WCDMA, changed to static final constants
- Changed concrete FrequencyBands and RfPort enums to only have the frequency/port letter respectively

- To update CarrierManagementSystemClient with updated expected test case results

### CODE REVIEW

File: src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java [hunk 1]
```diff
@@ -19,48 +19,47 @@
...
+    private final int LTE_MAXIMUM_CARRIER_ID = 500;
     private final int WCDMA_MAXIMUM_CARRIER_ID = 500;
     private Stack<Integer> lteCarrierIdGenerator = new Stack<>();
     private Stack<Integer> wcdmaCarrierIdGenerator = new Stack<>();
 
-	// For the future usage.
-	// This list will save all carrier object and carrierID so the object can be
-	// modified from radio unit.
-	// private ConcurrentHashMap<Integer,Carrier> carrierList;
+    // For the future usage.
```
Comment (john-breton) [inline src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java]: Since RadioUnits will have a list of Carriers, this comment here can probably be removed entirely.

File: src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java [hunk 2]
```diff
@@ -19,48 +19,47 @@
  * carrier it creates.
  */
 public class CarrierManagementSystemDirector implements CarrierManagementIf {
-    private static volatile CarrierManagementSystemDirector UNIQUE_INSTANCE = new CarrierManagementSystemDirector();
-    private final int LTE_MAXIMUM_CARRIER_ID = 500; 
+    private static final CarrierManagementSystemDirector UNIQUE_INSTANCE = new CarrierManagementSystemDirector();
```
Comment (john-breton) [inline src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java]: This should still be volatile (at least according to what Ista deems as the only "correct" Singleton). And the change should be in the getInstance() method.

File: src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java [hunk 3]
```diff
@@ -19,48 +19,47 @@
...
     }
-    
+
     /**
      * Get the unique CarrierManagementSystemDirector instance
      * 
      * @return CarrierManagementSystemDirector singleton instance
      */
-    public static CarrierManagementSystemDirector getInstance() {
+    public static synchronized CarrierManagementSystemDirector getInstance() {
```
Comment (john-breton) [inline src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java:53]: This method should be changed to something like the following: 
```
public static CarrierManagementSystemDirector getInstance() {
    if (UNIQUE_INSTANCE == null) {
        synchronized (CarrierManagementSystemDirector.class) {
            if (UNIQUE_INSTANCE == null) {
                UNIQUE_INSTANCE = new CarrierManagementSystemDirector();
            }
        }
    }
    return UNIQUE_INSTANCE;
}
```

File: src/main/java/common/Carrier.java [hunk 4]
```diff
@@ -10,7 +10,7 @@
  */
 public abstract class Carrier {
     protected final Integer carrierId;
-    protected final List<RfPort> rfPorts;
+    protected List<RfPort> rfPorts = null;
```
Comment (john-breton) [inline src/main/java/common/Carrier.java]: This doesn't need to be explicitly initialized to null here.

File: src/main/java/common/RatType.java [hunk 5]
```diff
@@ -5,8 +5,9 @@
  *
  */
 public enum RatType {
-	LTE("LTE"),
-	WCDMA("WCDMA");
+	
+    LTE("Long-Term Evolution"),
+	WCDMA("Wideband Code Division Multiple Access");
```
Comment (john-breton) [inline src/main/java/common/RatType.java:10]: When Ista assumed we all meant what WCDMA stood for I felt like I didn't belong as an Ericsson employee.

File: src/main/java/mediator/ConcreteMediator.java [hunk 6]
```diff
@@ -49,7 +49,7 @@ private ConcreteMediator() {
      *
      * @return The Singleton instance of the ConcreteMediator class.
      */
-    public static ConcreteMediator getInstance() {
+    public static synchronized ConcreteMediator getInstance() {
```
Comment (john-breton) [inline src/main/java/mediator/ConcreteMediator.java]: This will be getting updated in a separate commit. Basically exactly what I mentioned above for CarrierManagementSystemDirector.

**Review by john-breton — COMMENTED** [review]
These changes look good. The only additional thing I remember Ista mentioning was about combining the enums into a single file, but that can be done at a later time/not at all if we don't agree with it.
Some minor comments for now but I'll approve it regardless since everything I mentioned would be easy fixes.

File: src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java (hunk 3 above)
Comment (gabrielleungkh) [reply src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java:53]: I was talking with Ramin over this part with the double-checked locking, and since we chose eager instantiation as a way to keep the entire thing thread safe, I think just making getInstance() synchronized should be enough?

File: src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java (hunk 2 above)
Comment (gabrielleungkh) [reply src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java]: I've been trying to figure out how volatile vs synchronized should be used and I saw most times it was either one or the other, but I'm not sure enough of that so I'll just put it back :)

File: src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java (hunk 1 above)
Comment (gabrielleungkh) [reply src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java]: Done

File: src/main/java/common/Carrier.java (hunk 4 above)
Comment (gabrielleungkh) [reply src/main/java/common/Carrier.java]: When we were unit testing, it errored out without the explicit initialization, but I just tried again and it's working fine without so I'll take this out. Must have been something weird when we were still testing our change.

File: src/main/java/mediator/ConcreteMediator.java (hunk 6 above)
Comment (gabrielleungkh) [reply src/main/java/mediator/ConcreteMediator.java]: In that case I'll take this out and leave it for the separate commit from you to come in. I'd still like to discuss whether we should go forward with double checked locking over eager instantiation at some point though.

File: src/main/java/carriermanagementsystem/CarrierManagementIf.java [hunk 7]
```diff
@@ -9,35 +10,34 @@
...
+     * create LTE carrier based on inputs. If invalid parameters are passed in, no
+     * carrier will be created and the method will return null.
      *
-     * @param rfPorts            LTE carrier needs 4 ports. These RF ports can be
-     *                           selected from RFPorts enum class.
-     * @param frequencyBand LTE carrier needs specific band to operate. List of
-     *                           LTE carrier frequency are available in
-     *                           CarrierFrequencies enum class.
-     * @param transmittingPower  LTE carrier need transmitting power to be created.
+     * @param rfPorts           LTE carrier needs 4 ports. These RF ports can be
```
Comment (matthewkwsiu) [inline src/main/java/carriermanagementsystem/CarrierManagementIf.java:16]: Nice work on all the javadocs!

File: src/main/java/carriermanagementsystem/CarrierManagementSystemClient.java [hunk 8]
```diff
@@ -50,16 +50,16 @@ public static void main(String[] args) {
...
 //        /**
 //         * case #1: result:
 //         * 
 //         * Final result: 
-//         * LTE Carrier ID: 138
-//         * LTE Carrier Frequency: LTE_BAND_4 = 1755 MHz
-//         * LTE RF Ports: RF_3 = D  RF_4 = E  RF_5 = F  RF_6 = G  
+//         * LTE Carrier ID: 15
+//         * LTE Carrier Frequency: 1710 MHz
+//         * LTE RF Ports: D  E  F  G
```
Comment (matthewkwsiu) [inline src/main/java/carriermanagementsystem/CarrierManagementSystemClient.java]: These were to address his points on just using letters right?

File: src/main/java/carriermanagementsystem/CarrierManagementSystemClient.java [hunk 9]
```diff
@@ -68,27 +68,38 @@ public static void main(String[] args) {
 //         * case #2: pass null for the RfPorts 
 //         */
 //        
-//        FrequencyBand WcdmaBand_6 = FrequencyBand.WCDMA_BAND_6;
+//        FrequencyBand WcdmaBand_5 = FrequencyBand.WCDMA_BAND_5;
```
Comment (matthewkwsiu) [inline src/main/java/carriermanagementsystem/CarrierManagementSystemClient.java:74]: Pardon, to quickly clarify, was there a reason for changing these bands in response to a point of feedback Ista had gave? 

Or was this for testing?

File: src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java [hunk 10]
```diff
@@ -20,47 +20,42 @@
  */
 public class CarrierManagementSystemDirector implements CarrierManagementIf {
     private static volatile CarrierManagementSystemDirector UNIQUE_INSTANCE = new CarrierManagementSystemDirector();
-    private final int LTE_MAXIMUM_CARRIER_ID = 500; 
+    private final int LTE_MAXIMUM_CARRIER_ID = 500;
```
Comment (matthewkwsiu) [inline src/main/java/carriermanagementsystem/CarrierManagementSystemDirector.java:23]: Nice whitespace changes. Was this from CTRL + SHIFT + F?

File: src/main/java/carriermanagementsystem/WcdmaCarrierBuilder.java [hunk 11]
```diff
@@ -8,9 +8,11 @@
 import java.util.List;
 
 /**
- * This class is a child class to CarrierBuilder. It builds WCDMACarrier objects.
+ * This class is a child class to CarrierBuilder. It builds WCDMACarrier
+ * objects.
  */
 public class WcdmaCarrierBuilder implements CarrierBuilder {
+    private static final int WCDMA_RF_PORTS_NUMBER = 2;
```
Comment (matthewkwsiu) [inline src/main/java/carriermanagementsystem/WcdmaCarrierBuilder.java:16]: I appreciate you adding this variable to remove the magic number (just putting '2' in the code) from this class. Nice work.

File: src/main/java/userinterface/UserInterface.java [hunk 12]
```diff
@@ -144,25 +146,25 @@ public static void main(String[] args) {
...
-        RF_PORT_NUMBER(int value) {
-            this.value = value;
-        }
-
-        public int getValue() {
-            return value;
-        }
-
-    }
+//    /**
```
Comment (matthewkwsiu) [inline src/main/java/userinterface/UserInterface.java]: To check, are you keeping these as commented out since you'd like to keep them as reference, or should these be removed?

**Review by matthewkwsiu — COMMENTED** [review]
Thanks for addressing the changes that Ista had done. I've left some comments and questions. Nice work!

One small note: I appreciate that you adjusted formatting and spacing in some of these files (I assume you used CTRL + SHIFT + F) but it can sometimes make it a little more challenging to review the commit because of how all the whitespace changes are treated as individual changes and thus crowds out some of the practical changes you had done. This was feedback I had received in the past. I'm not sure if applies very much here, especially as this is a casual project, but I thought I'd share the feedback with you.

File: src/main/java/carriermanagementsystem/CarrierManagementSystemClient.java (hunk 9 above)
Comment (gabrielleungkh) [reply src/main/java/carriermanagementsystem/CarrierManagementSystemClient.java:74]: I realized the variable names just didn't match up with the examples so I changed them back

**Review by john-breton — APPROVED** [review]
Everything looks good to me, I don't think we need to delay this getting merged any longer.
